### PR TITLE
fix(session): limit max active sessions and calculate session ttl serverside

### LINKED ISSUE #98: Session duration

### PR DESCRIPTION
**Pull request by MikMik1011**
Closes #98 and #102

### CODE REVIEW

**Comment by aleksasiriski** [comment]
Refactor doesn't affect code logic, if you changed the code logic (or db schema) it should be a fix

File: src/lib/server/db/session.ts [hunk 1]
```diff
@@ -64,16 +69,15 @@ export async function validateSessionToken(token: string): Promise<SessionValida
...
 		const { user, session } = result[0];
-		if (Date.now() >= session.expiresAt.getTime()) {
+		if (Date.now() > session.timestamp.getTime() + INACTIVITY_TIMEOUT) {
 			await db.delete(sessionTable).where(eq(sessionTable.id, session.id));
 			return { session: null, user: null };
-		}
-		if (Date.now() >= session.expiresAt.getTime() - 1000 * 60 * 60 * 24 * 15) {
-			session.expiresAt = new Date(Date.now() + 1000 * 60 * 60 * 24 * 30);
+		} else {
+			session.timestamp = new Date(Date.now());
```
Comment (aleksasiriski) [inline src/lib/server/db/session.ts:76]: Use just Date.now()

Comment (MikMik1011) [reply src/lib/server/db/session.ts:76]: `Date.now()` returns just a number and in our database `timestamp` is a date, however, I agree, this line is bad so i will just use `new Date()`.

File: src/routes/login/+page.server.ts [hunk 2]
```diff
@@ -58,7 +58,10 @@ export const actions: Actions = {
 			// Create a new session token
 			const sessionToken = generateSessionToken();
 			const session = await createSession(sessionToken, id, building);
-			setSessionTokenCookie(event, sessionToken, session.expiresAt);
+			setSessionTokenCookie(event, sessionToken, session.timestamp);
+
+			// Invalidate sessions that exceed the maximum number of sessions
+			invalidateExcessSessions(id);
```
Comment (aleksasiriski) [inline src/routes/login/+page.server.ts:64]: This is async, it must be awaited

File: src/lib/server/db/session.ts [hunk 3]
```diff
@@ -1,9 +1,14 @@
...
+import { and, desc, eq, notInArray } from 'drizzle-orm';
 import { encodeBase32LowerCaseNoPadding, encodeHexLowerCase } from '@oslojs/encoding';
 import { sha256 } from '@oslojs/crypto/sha2';
 import { userTable, type User } from './schema/user';
 import { sessionTable, type Session } from './schema/session';
 import { DB as db } from './connect';
+import { env } from '$env/dynamic/private';
+
+
+export const INACTIVITY_TIMEOUT = Number.parseInt(env.INACTIVITY_TIMEOUT ?? '30') * 60 * 1000;
```
Comment (aleksasiriski) [inline src/lib/server/db/session.ts:10]: Why uppersnakecase? Just use camel case

File: src/lib/server/db/session.ts [hunk 4]
```diff
@@ -64,16 +69,15 @@ export async function validateSessionToken(token: string): Promise<SessionValida
 			return { session: null, user: null };
 		}
 		const { user, session } = result[0];
-		if (Date.now() >= session.expiresAt.getTime()) {
+		if (Date.now() > session.timestamp.getTime() + INACTIVITY_TIMEOUT) {
```
Comment (aleksasiriski) [inline src/lib/server/db/session.ts:72]: >=

File: src/lib/server/db/session.ts [hunk 5]
```diff
@@ -96,6 +100,31 @@ export async function invalidateSession(sessionId: string): Promise<void> {
 	}
 }
 
+// Invalidate sessions that exceed the maximum number of sessions
+export async function invalidateExcessSessions(userId: number): Promise<number> {
```
Comment (aleksasiriski) [inline src/lib/server/db/session.ts:104]: This select and delete should happen within a transaction

Comment (aleksasiriski) [inline src/lib/server/db/session.ts:104]: Maybe put this select and delete within a transaction? @matijakljajic your opinion on this pls

File: src/lib/server/db/session.ts [hunk 6]
```diff
@@ -96,6 +100,31 @@ export async function invalidateSession(sessionId: string): Promise<void> {
...
+			id: sessionTable.id
+		})
+		.from(sessionTable)
+		.where(eq(sessionTable.userId, userId))
+		.orderBy(desc(sessionTable.timestamp))
+		.limit(MAX_ACTIVE_SESSIONS);
+
+	const sessionIdsToKeep = newestSessions.map((session) => session.id);
+
+	const deletedSessions = await db
```
Comment (aleksasiriski) [inline src/lib/server/db/session.ts:121]: w/o `returning()` the deletedSessions is of type "never[]", but also why would we even want to know this info? I would just change this function to return void and delete this const deletedSessions

**Review by aleksasiriski — CHANGES_REQUESTED** [review]
Great work, just a few more changes before merging :) Also, remember to squash merge this since your commit messages aren't following semantic release (refactor shouldn't change code logic)

**Pull request merged**